tokio/src/sync/watch.rs

```diff
@@ -266,79 +228,110 @@ impl<T: Clone> Receiver<T> {
     /// async fn main() {
     ///     let (tx, mut rx) = watch::channel("hello");
     ///
-    ///     let v = rx.recv().await;
-    ///     assert_eq!(v, "hello");
-    ///
     ///     tokio::spawn(async move {
     ///         tx.send("goodbye").unwrap();
     ///     });
     ///
-    ///     // Waits for the new task to spawn and send the value.
-    ///     let v = rx.recv().await;
-    ///     assert_eq!(v, "goodbye");
+    ///     assert!(rx.changed().await.is_ok());
+    ///     assert_eq!(*rx.borrow(), "goodbye");
     ///
-    ///     let v = rx.recv().await;
-    ///     assert_eq!(v, "goodbye");
+    ///     // The `tx` handle has been dropped
+    ///     assert!(rx.changed().await.is_err());
     /// }
     /// ```
-    pub async fn recv(&mut self) -> T {
-        poll_fn(|cx| {
-            let v_ref = ready!(self.poll_recv_ref(cx));
-            Poll::Ready((*v_ref).clone())
+    pub async fn changed(&mut self) -> Result<(), error::RecvError> {
+        use std::future::Future;
+        use std::pin::Pin;
+        use std::task::Poll;
+
+        // In order to avoid a race condition, we first request a notification,
+        // **then** check the current value's version. If a new version exists,
+        // the notification request is dropped. Requesting the notification
+        // requires polling the future once.
+        let notified = self.shared.notify_rx.notified();
+        pin!(notified);
+
+        // Polling the future once is guaranteed to return `Pending` as `watch`
+        // only notifies using `notify_waiters`.
+        crate::future::poll_fn(|cx| {
```

Comment:
This forces a waiter to be pushed by `Notify`. I thought about moving it into a fn, but I opted to punt.

Reply:
Yup, this seems fine for now!